src/hotspot/share/runtime/crac.cpp

```diff
@@ -602,3 +656,1075 @@ void crac::update_javaTimeNanos_offset() {
+                            vmClasses::jdk_internal_loader_ClassLoaders_klass()->external_name(), dump->id));
+        }
+        if (field_dump.value.as_object_id == NULL_ID) {
+          THROW_MSG(vmSymbols::java_lang_IllegalArgumentException(), "Dumped built-in app class loader is null");
+        }
+        _builtin_app_class_loader_id = field_dump.value.as_object_id;
+      }
+    }
+
+    if (_platform_class_loader_id == NULL_ID || _builtin_app_class_loader_id == NULL_ID) {
+      ResourceMark rm;
+      THROW_MSG(vmSymbols::java_lang_IllegalArgumentException(),
+                err_msg("Static field %s and/or %s are missing from %s dump " INT64_FORMAT,
+                        PLATFORM_LOADER_FIELD_NAME, APP_LOADER_FIELD_NAME, vmClasses::ClassLoader_klass()->external_name(),
+                        dump->id));
+    }
+  }
+
+  ResizeableResourceHashtable<HeapDumpFormat::id_t, instanceHandle, AnyObj::C_HEAP> _prepared_class_loaders {11,   1228891};
+  ResizeableResourceHashtable<HeapDumpFormat::id_t, instanceHandle, AnyObj::C_HEAP> _allocated_prot_domains {11,   1228891};
+  ResizeableResourceHashtable<HeapDumpFormat::id_t, Klass *,        AnyObj::C_HEAP> _loaded_classes         {107,  1228891};
+  ResizeableResourceHashtable<HeapDumpFormat::id_t, Klass *,        AnyObj::C_HEAP> _restored_classes       {107,  1228891};
+  ResizeableResourceHashtable<HeapDumpFormat::id_t, Handle,         AnyObj::C_HEAP> _restored_objects       {1009, 1228891};
+
+  // Gets a value or puts an empty-initialized value if the key is absent.
+  template <class V>
+  V *get_or_put_stub(HeapDumpFormat::id_t id, ResizeableResourceHashtable<HeapDumpFormat::id_t, V, AnyObj::C_HEAP> *table) {
+    V *value = table->get(id);
+    if (value == nullptr) {
+      table->put_when_absent(id, {});
+      table->maybe_grow();
+    }
+    return value;
+  }
+
+  // Loads the class without restoring it.
+  Klass *load_class(const HeapDumpFormat::ClassDumpRecord &dump, TRAPS) {
+    Klass **ready = get_or_put_stub(dump.id, &_loaded_classes);
+    if (ready != nullptr) {
+      if (*ready == nullptr) {
+        THROW_MSG_NULL(vmSymbols::java_lang_IllegalArgumentException(),
+                       err_msg("Loading curcularity detected for dumped class " INT64_FORMAT, dump.id));
+      }
+      return *ready;
+    }
+    log_trace(restore)("Loading class " INT64_FORMAT, dump.id);
+
+    // TODO have to also load interfaces, because otherwise the standard
+    // resolution mechanism will be called for them which may call the user code
+    if (dump.super_id != NULL_ID) {
+      HeapDumpFormat::ClassDumpRecord *super_dump = get_class_dump(dump.super_id, CHECK_NULL);
+      load_class(*super_dump, CHECK_NULL);
+    }
+
+    Symbol *name = get_dumped_class_name(dump.id, CHECK_NULL);
+
+    instanceHandle class_loader = get_prepared_class_loader(dump.class_loader_id, CHECK_NULL);
```

Comment:
Some general questions about handles (not about this particular place in the code):
1. Do I understand right that I need to use handles in the native VM code only when there is a possibility of a safepoint (because only then a GC can happen concurrently)? Is code like the following always safe?
```c++
oop obj = get_oop();  // some function that returns an oop
// <-- how do I know a GC won't screw up that oop here, before I can even try to put it into a handle?
Handle h(Thread::current(), oop);
```
2. There is this thing called `HandleMark` -- when do I need to use one? I pretty much understand what it does, but its docs say "It should normally not be necessary to use HandleMarks manually". I have to use `ResourceMark`s every time I allocate a resource, shouldn't I also use `HandleMark`s when I allocate handles?

Reply:
1. Yes, safepoints in VM code are explicit, so the code is safe.

2. I assume you  don't have to use HandleMark. It would make sense if you'd be allocating a lot of handles and would like to free them sooner (you probably will know, when you're such kind of trouble). But unlikely that is the case.

```//------------------------------------------------------------------------------------------------------------------------
// Handles are allocated in a (growable) thread local handle area. Deallocation
// is managed using a HandleMark. It should normally not be necessary to use
// HandleMarks manually.
//
// A HandleMark constructor will record the current handle area top, and the
// destructor will reset the top, destroying all handles allocated in between.
// The following code will therefore NOT work:
//
//   Handle h;
//   {
//     HandleMark hm(THREAD);
//     h = Handle(THREAD, obj);
//   }
//   h()->print();       // WRONG, h destroyed by HandleMark destructor.
//
// If h has to be preserved, it can be converted to an oop or a local JNI handle
// across the HandleMark boundary.

Reply:
ResourceMark's very often clears a lot of strings of arbitrary lengths, so they make more sense in your code and making them in call stacks as deep as possible